wdl-engine/src/eval/v1.rs

```diff
@@ -0,0 +1,2984 @@
+
+        let value = eval_v1_expr(
+            V1::Zero,
+            r#"if (foo) then baz else "str""#,
+            &mut types,
+            scope,
+        )
+        .unwrap();
+        assert_eq!(value.unwrap_file().as_str(), "file");
+
+        let value = eval_v1_expr(
+            V1::Zero,
+            r#"if (bar) then baz else "path""#,
+            &mut types,
+            scope,
+        )
+        .unwrap();
+        assert_eq!(value.unwrap_file().as_str(), "path");
+    }
+
+    #[test]
+    fn logical_not_expr() {
+        let scopes = &[Scope::new(None)];
+        let scope = ScopeRef::new(scopes, 0);
+
+        let mut types = Types::default();
+        let value = eval_v1_expr(V1::Zero, r#"!true"#, &mut types, scope).unwrap();
+        assert!(!value.unwrap_boolean());
+
+        let value = eval_v1_expr(V1::Zero, r#"!false"#, &mut types, scope).unwrap();
+        assert!(value.unwrap_boolean());
+    }
+
+    #[test]
+    fn negation_expr() {
+        let scopes = &[Scope::new(None)];
+        let scope = ScopeRef::new(scopes, 0);
+
+        let mut types = Types::default();
+        let value = eval_v1_expr(V1::Zero, r#"-1234"#, &mut types, scope).unwrap();
+        assert_eq!(value.unwrap_integer(), -1234);
+
+        let value = eval_v1_expr(V1::Zero, r#"-(1234)"#, &mut types, scope).unwrap();
+        assert_eq!(value.unwrap_integer(), -1234);
+
+        let value = eval_v1_expr(V1::Zero, r#"----1234"#, &mut types, scope).unwrap();
+        assert_eq!(value.unwrap_integer(), 1234);
+
+        let value = eval_v1_expr(V1::Zero, r#"-1234.5678"#, &mut types, scope).unwrap();
+        approx::assert_relative_eq!(value.unwrap_float(), -1234.5678);
+
+        let value = eval_v1_expr(V1::Zero, r#"-(1234.5678)"#, &mut types, scope).unwrap();
+        approx::assert_relative_eq!(value.unwrap_float(), -1234.5678);
+
+        let value = eval_v1_expr(V1::Zero, r#"----1234.5678"#, &mut types, scope).unwrap();
+        approx::assert_relative_eq!(value.unwrap_float(), 1234.5678);
+    }
+
+    #[test]
+    fn logical_or_expr() {
+        let scopes = &[Scope::new(None)];
+        let scope = ScopeRef::new(scopes, 0);
+
+        let mut types = Types::default();
+        let value = eval_v1_expr(V1::Zero, r#"false || false"#, &mut types, scope).unwrap();
+        assert!(!value.unwrap_boolean());
+
+        let value = eval_v1_expr(V1::Zero, r#"false || true"#, &mut types, scope).unwrap();
+        assert!(value.unwrap_boolean());
+
+        let value = eval_v1_expr(V1::Zero, r#"true || false"#, &mut types, scope).unwrap();
+        assert!(value.unwrap_boolean());
+
+        let value = eval_v1_expr(V1::Zero, r#"true || true"#, &mut types, scope).unwrap();
+        assert!(value.unwrap_boolean());
+
+        let value = eval_v1_expr(V1::Zero, r#"true || nope"#, &mut types, scope).unwrap();
+        assert!(value.unwrap_boolean());
+
+        let diagnostic = eval_v1_expr(V1::Zero, r#"false || nope"#, &mut types, scope).unwrap_err();
```

Comment:
Though it's cool that something like this is left to runtime, I do wonder if it needs to be statically checked to avoid errors (possibly a new lint rule?).

Reply:
`true || nope` (where `nope` is an unknown name, like in this case) or where `nope` might be a name that resolves to a non-boolean type are caught at static analysis time.

The expression evaluator, though, is implemented in such a way that it doesn't care what the RHS is.

When we hook this all up to `sprocket`, we'll be reporting on any error-level static analysis diagnostics and exiting non-zero before we even get to evaluation.

Reply:
Okay, great! Then that is taken care of.